_posts/2017-07-23-fastlane-travis-weekly-deploys.md

```diff
@@ -0,0 +1,314 @@
+
+```yml
+# Use a Mac build please
+language: objective-c
+osx_image: xcode8.2
+
+# Ensure that Fastlane is at the latest version
+before_install:
+- bundle update
+
+# Let fastlane set up the other dependency managers
+before_script:
+- bundle exec fastlane setup
+
+# Separate fastlane lanes so that they can be individually
+# tested one by one during development
+script:
+- bundle exec fastlane validate_env_vars
+- bundle exec fastlane stamp_plist
+- bundle exec fastlane ship
+```
+
+Note the `- bundle update`. As fastlane works against iTunes connect which is always changing locking the version doesn't make much sense. 
+
+## Ensuring signing will work.
+
+This one is a bit tricker, luckily I've already set up one of our apps to use [fastlane match][] and I can re-use that infrastructure. As it is a private repo, Travis did not have access to clone the repo. I fixed this by creating an access token for a user with read-only access to our match-codesigning repo, then exposed this as a private environment variable in CI which the Matchfile uses. E.g.
+
+```ruby
+git_url "https://#{ENV['GITHUB_SUBMODULES_USER']}@github.com/artsy/mobile_fastlane_match"
+
+# Instead of 
+# git_url "https://github.com/artsy/mobile_fastlane_match"
+```
+
+This is one of the highlights on fastlane's choice in building a DSL that which sits above a real programming language, you give users a lot of flexibility.
+
+Next up, I added a fastlane lane for code signing, and keychain setup. This just calls two setup functions.
+
+```ruby
+lane :setup_signing do
+  create_keychain(
+    name: ENV['MATCH_KEYCHAIN_NAME'],
+    password: ENV['MATCH_PASSWORD'],
+    default_keychain: true,
+    unlock: true,
+    timeout: 3600,
+    add_to_search_list: true
+  )
+
+  match(type: 'appstore',
+        keychain_name: ENV['MATCH_KEYCHAIN_NAME'],
+        keychain_password: ENV['MATCH_PASSWORD'],
+        readonly: true)
+end
+```
+
+## Creating the build and shipping it to Testflight
+
+This is handled by [fastlane gym][gym] at the start of the main lane.
+
+```ruby
+# You can run `bundle exec fastlane ship` to test the build process locally.
+lane :ship do
+  # Run the code signing lane
+  setup_signing
+
+  # We were having issues with building an a few folders deep.
+  # The /Pods bit is because we can rely on it being there, see
+  # this link: https://docs.fastlane.tools/advanced/#directory-behavior
+  Dir.chdir('../emission/Example/Pods') do
+    gym workspace: 'Emission.xcworkspace',
+        configuration: 'Deploy',
+        scheme: 'Emission'
+  end
+
+  # [...]
+end
+```
+
+It uses a scheme for deploys, which prioritises using AppHub over a local React Native server. Gym handles a lot of CLI ugliness for us, and works well.
+
+Sending the app to Testflight involves a little bit of work:
```

Comment:
`Preparing a changelog for TestFlight requires us to do some manual parsing`

The TestFlight piece is just one line of code :)